cunumeric/array.py

```diff
@@ -2040,6 +2067,47 @@ def diagonal(
                 raise ValueError("Either axis1/axis2 or axes must be supplied")
         return self._diag_helper(offset=offset, axes=axes, extract=extract)
 
+    def trace(self, offset=0, axis1=None, axis2=None, dtype=None, out=None):
+        """a.trace(offset=0, axis1=None, axis2=None, dtype = None, out = None)
+
+        Return the sum along diagonals of the array.
+
+        Refer to :func:`cunumeric.trace` for full documentation.
+
+        See Also
+        --------
+        cunumeric.trace : equivalent function
+
+        Availability
+        --------
+        Multiple GPUs, Multiple CPUs
+
+        """
+        if self.ndim < 2:
+            raise ValueError(
+                "trace operation can't be called on a array with DIM<2"
+            )
+
+        axes = []
+        if (axis1 is None) and (axis2 is None):
+            # default values for axis
+            axes = (0, 1)
+        elif (axis1 is None) or (axis2 is None):
+            raise ValueError("both axes should be passed")
+        else:
+            axes = (axis1, axis2)
+
+        res = self._diag_helper(offset=offset, axes=axes, trace=True, out=out)
+
+        # for 2D arrays we must return scalar
+        if self.ndim == 2:
+            res = res[0]
+
+        if out is None and dtype is not None:
+            res = res.astype(dtype)
```

Comment:
`dtype` should not be used to cast-out the result of the computation, it should be passed in to `_diag_helper` to set the type used in the computation (i.e. `a` should be cast to this dtype if present, unless `out` is also present, in which case `out`'s dtype wins).

Reply:
can I still use `astype` for this?